Source/WebCore/platform/graphics/ArrayPixelBuffer.h

```diff
@@ -0,0 +1,179 @@
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "PixelBuffer.h"
+#include <JavaScriptCore/ArrayBufferView.h>
+#include <JavaScriptCore/TypedArrayAdaptersForwardDeclarations.h>
+#include <JavaScriptCore/TypedArrayInlines.h>
+
+namespace WebCore {
+
+class ArrayPixelBuffer : public PixelBuffer {
+public:
+    WEBCORE_EXPORT static std::optional<Ref<ArrayPixelBuffer>> create(const PixelBufferFormat&, const IntSize&, std::span<const uint8_t> data);
+
+    virtual ~ArrayPixelBuffer() = default;
+
+    JSC::ArrayBufferView& data() const LIFETIME_BOUND { return m_data.get(); }
+    Ref<JSC::ArrayBufferView> protectedData() const { return m_data; }
+    Ref<JSC::ArrayBufferView>&& takeData() && { return WTFMove(m_data); }
+    std::span<const uint8_t> span() const LIFETIME_BOUND { return m_data->span(); }
+
+protected:
+    ArrayPixelBuffer(const PixelBufferFormat&, const IntSize&, Ref<JSC::ArrayBufferView>&&);
+
+private:
+    Ref<JSC::ArrayBufferView> m_data;
+};
+
+template <typename JSCTypedArrayView, PixelBuffer::Type pixelBufferType, auto... supportedPixelFormats>
+class TypedArrayPixelBuffer final : public ArrayPixelBuffer {
+public:
+    WEBCORE_EXPORT static Ref<TypedArrayPixelBuffer> create(const PixelBufferFormat&, const IntSize&, JSCTypedArrayView&);
+    WEBCORE_EXPORT static std::optional<Ref<TypedArrayPixelBuffer>> create(const PixelBufferFormat&, const IntSize&, std::span<const uint8_t> data);
+
+    WEBCORE_EXPORT static RefPtr<TypedArrayPixelBuffer> tryCreate(const PixelBufferFormat&, const IntSize&);
+    WEBCORE_EXPORT static RefPtr<TypedArrayPixelBuffer> tryCreate(const PixelBufferFormat&, const IntSize&, Ref<JSC::ArrayBuffer>&&);
+
+    JSCTypedArrayView& data() const LIFETIME_BOUND { return uncheckedDowncast<JSCTypedArrayView>(ArrayPixelBuffer::data()); }
+    Ref<JSCTypedArrayView> protectedData() const { return data(); }
+    Ref<JSCTypedArrayView> takeData() && { return adoptRef(uncheckedDowncast<JSCTypedArrayView>(WTFMove(*this).ArrayPixelBuffer::takeData().leakRef())); }
+
+    Type type() const override { return pixelBufferType; }
+    RefPtr<PixelBuffer> createScratchPixelBuffer(const IntSize&) const override;
+
+private:
+    TypedArrayPixelBuffer(const PixelBufferFormat&, const IntSize&, Ref<JSCTypedArrayView>&&);
+};
+
+template <typename JSCTypedArrayView, PixelBuffer::Type type, auto... supportedPixelFormats>
```

Comment:
Please move the implementation to a cpp file. Many engineers in WebKit do not like the template class implementation in the header file. I think all you need to add at the end of this source file is:
```
// Explicit template instantiation
template class TypedArrayPixelBuffer<JSC::Uint8ClampedArray, ...>;
template class TypedArrayPixelBuffer<JSC::Float16Array, ...>;
```